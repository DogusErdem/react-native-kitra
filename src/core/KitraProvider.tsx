import type { ComponentType, ForwardedRef } from 'react';
import React from 'react';
import { GestureHandlerRootView } from 'react-native-gesture-handler';
import Animated from 'react-native-reanimated';
import { StyleProp, ViewStyle } from 'react-native';

import { ThemeProvider } from './theme/theme';
import { TypographyContext, TypographyProvider } from './typography/typography';
import { NotificationContextType, NotificationProvider } from './notification/notification';

const Feather = require('react-native-vector-icons/Feather').default;
const Ionicons = require('react-native-vector-icons/Ionicons').default;
const AntDesign = require('react-native-vector-icons/AntDesign').default;

Animated.addWhitelistedNativeProps({ text: true });
const messageTypes = (theme:any) => ({
  SUCCESS: {
    backgroundColor: theme.colors.status.successLight,
    icon: <Feather name="check" size={25} color={theme.colors.status.success} />,
  },
  INFO: {
    backgroundColor: theme.colors.system.primary15,
    icon: <Ionicons name="information-circle-outline" size={25} color={theme.colors.system.primary} />,
  },
  DANGER: {
    backgroundColor: theme.colors.status.errorLight,
    icon: <AntDesign name="warning" size={25} color={theme.colors.status.error} />,
  },
  WARNING: {
    backgroundColor: theme.colors.status.warningLight,
    icon: <AntDesign name="warning" size={25} color={theme.colors.status.warning} />,
  },
});

type KitraProviderType= {
  children?:any,
  limit?:number,
  notificationRef?:React.Ref<NotificationContextType>,
  messageType?:(theme?:any, colorScheme?:'dark' | 'light') => {[key:string]: {
      backgroundColor:string,
    icon?:React.ReactNode
    onPress?:()=>void
  }},
  notificationCcontainerStyle?:StyleProp<ViewStyle>,
  customView?:({ header, type, message, theme }:
    {header:string, type:string, message:string, theme:any})=>React.ReactNode}

export const KitraProvider =
({ children, messageType = messageTypes, notificationCcontainerStyle, limit, customView,
  notificationRef }: KitraProviderType) => (
    <GestureHandlerRootView style={{ flex: 1 }}>
      <ThemeProvider>
        <TypographyProvider>
          <NotificationProvider
            messageType={theme => (messageType ? messageType(theme) : theme)}
            notificationCcontainerStyle={notificationCcontainerStyle}
            limit={limit}
            customView={item => customView?.(item)}
            ref={notificationRef}
          >
            {children}
          </NotificationProvider>
        </TypographyProvider>
      </ThemeProvider>
    </GestureHandlerRootView>
);

export function applyDefaults<T>(Component: ComponentType<T>) {
  return React.forwardRef((props: T, ref: ForwardedRef<any>) => (
    <TypographyContext.Consumer>
      {typography =>
        <Component ref={ref} typography={typography?.typography} {...props} />
          }
    </TypographyContext.Consumer>
  ));
}
