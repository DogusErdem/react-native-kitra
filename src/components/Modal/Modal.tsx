// @ts-nocheck

import { useState, useEffect } from 'react';
import { View, Modal as RNModal, StyleSheet, ViewStyle, ModalProps as RNModalProps } from 'react-native';
import type { FCCWD, ModalProps } from '../../types';
import { opacity } from '../../utilities';
import { applyDefaults } from '../../core/KitraProvider';

const Modal: FCCWD<ModalProps&RNModalProps> = ({ theme, containerStyle, modalStyle, visible = false, children, ...props }) => {
  const [modalVisible, setModalVisible] = useState(visible);
  useEffect(() => {
    setModalVisible(visible);
  }, [visible]);
  return (
    <View style={{ flex: 1 }}>
      <RNModal
        animationType="slide"
        transparent
        visible={modalVisible}
        {...props}
      >
        <View style={[styles.centeredView, { backgroundColor: opacity(String(theme?.lightBlack), 80) }, containerStyle]}>
          <View style={[styles.modalView, modalStyle]}>
            {children}
          </View>
        </View>
      </RNModal>
    </View>
  );
};

const styles = StyleSheet.create({
  centeredView: {
    justifyContent: 'center',
    alignItems: 'center',
  },
  modalView: {
  },
});

export default applyDefaults(Modal);
