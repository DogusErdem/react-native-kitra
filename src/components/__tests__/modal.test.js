/* eslint-disable no-undef */
import { fireEvent, render, act, screen } from '@testing-library/react-native';
import { KitraProvider } from '../../core/KitraProvider';
import Modal from '../Modal/Modal';

const renderWithContext = children => render(<KitraProvider>{children}</KitraProvider>);

describe('Modal', () => {
  it('renders Modal visible is false', () => {
    const tree = renderWithContext(<Modal />);
    expect(tree).toMatchSnapshot();
  });
  it('renders Modal visible is true', () => {
    const tree = renderWithContext(<Modal visible />);
    expect(tree).toMatchSnapshot();
  });
});
