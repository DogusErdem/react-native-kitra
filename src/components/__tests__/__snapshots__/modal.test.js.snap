// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders Modal visible is false 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <Modal
      animationType="slide"
      hardwareAccelerated={false}
      transparent={true}
      typography={
        {
          "body": {
            "medium": {
              "fontFamily": "Poppins-Medium",
              "fontSize": 14,
              "lineHeight": 21,
            },
            "regular": {
              "fontFamily": "Poppins-Regular",
              "fontSize": 14,
              "lineHeight": 16,
            },
            "smedium": {
              "fontFamily": "Poppins-Medium",
              "fontSize": 12,
              "lineHeight": 18,
            },
            "sregular": {
              "fontFamily": "Poppins-Regular",
              "fontSize": 12,
              "lineHeight": 16,
            },
            "xsmedium": {
              "fontFamily": "Poppins-Medium",
              "fontSize": 10,
              "lineHeight": 16,
            },
            "xsregular": {
              "fontFamily": "Poppins-Regular",
              "fontSize": 10,
              "lineHeight": 14,
            },
            "xxsmedium": {
              "fontFamily": "Poppins-Medium",
              "fontSize": 8,
              "lineHeight": 12,
            },
            "xxsregular": {
              "fontFamily": "Poppins-Regular",
              "fontSize": 8,
              "lineHeight": 12,
            },
          },
          "heading": {
            "h1": {
              "fontFamily": "Poppins-Medium",
              "fontSize": 30,
              "lineHeight": 45,
            },
            "h2": {
              "fontFamily": "Poppins-Medium",
              "fontSize": 28,
              "lineHeight": 42,
            },
            "h3": {
              "fontFamily": "Poppins-Medium",
              "fontSize": 23,
              "lineHeight": 35,
            },
            "h4": {
              "fontFamily": "Poppins-Medium",
              "fontSize": 19,
              "lineHeight": 29,
            },
            "h5": {
              "fontFamily": "Poppins-Medium",
              "fontSize": 16,
              "lineHeight": 25,
            },
          },
        }
      }
      visible={false}
    />
  </View>
  <View
    entering={BaseAnimationMock {}}
    exiting={BaseAnimationMock {}}
    layout={BaseAnimationMock {}}
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 10,
          "height": 100,
          "paddingHorizontal": 24,
          "width": "100%",
          "zIndex": 100,
        },
        {
          "alignSelf": "center",
          "top": 150,
        },
        undefined,
        {
          "backgroundColor": "transparent",
          "marginTop": -110,
          "position": "absolute",
        },
      ]
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "flex": 1,
          "flexDirection": "row",
          "opacity": 1,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 100,
            "height": 28,
            "justifyContent": "center",
            "marginLeft": 18,
            "marginTop": 21,
            "width": 28,
            "zIndex": 100,
          }
        }
      />
      <View
        style={
          {
            "flex": 1,
            "zIndex": 100,
          }
        }
      >
        <Text
          ellipsizeMode="middle"
          numberOfLines={3}
          style={
            [
              {
                "fontWeight": "500",
                "marginTop": 17,
                "paddingHorizontal": 20,
              },
              {
                "color": "#FFFFFF",
              },
            ]
          }
        />
        <Text
          ellipsizeMode="middle"
          numberOfLines={3}
          style={
            [
              {
                "fontWeight": "400",
                "marginTop": 6,
                "paddingHorizontal": 20,
              },
              {
                "color": "#FFFFFF",
              },
            ]
          }
        />
      </View>
    </View>
  </View>
</View>
`;

exports[`Modal renders Modal visible is true 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <Modal
      animationType="slide"
      hardwareAccelerated={false}
      transparent={true}
      typography={
        {
          "body": {
            "medium": {
              "fontFamily": "Poppins-Medium",
              "fontSize": 14,
              "lineHeight": 21,
            },
            "regular": {
              "fontFamily": "Poppins-Regular",
              "fontSize": 14,
              "lineHeight": 16,
            },
            "smedium": {
              "fontFamily": "Poppins-Medium",
              "fontSize": 12,
              "lineHeight": 18,
            },
            "sregular": {
              "fontFamily": "Poppins-Regular",
              "fontSize": 12,
              "lineHeight": 16,
            },
            "xsmedium": {
              "fontFamily": "Poppins-Medium",
              "fontSize": 10,
              "lineHeight": 16,
            },
            "xsregular": {
              "fontFamily": "Poppins-Regular",
              "fontSize": 10,
              "lineHeight": 14,
            },
            "xxsmedium": {
              "fontFamily": "Poppins-Medium",
              "fontSize": 8,
              "lineHeight": 12,
            },
            "xxsregular": {
              "fontFamily": "Poppins-Regular",
              "fontSize": 8,
              "lineHeight": 12,
            },
          },
          "heading": {
            "h1": {
              "fontFamily": "Poppins-Medium",
              "fontSize": 30,
              "lineHeight": 45,
            },
            "h2": {
              "fontFamily": "Poppins-Medium",
              "fontSize": 28,
              "lineHeight": 42,
            },
            "h3": {
              "fontFamily": "Poppins-Medium",
              "fontSize": 23,
              "lineHeight": 35,
            },
            "h4": {
              "fontFamily": "Poppins-Medium",
              "fontSize": 19,
              "lineHeight": 29,
            },
            "h5": {
              "fontFamily": "Poppins-Medium",
              "fontSize": 16,
              "lineHeight": 25,
            },
          },
        }
      }
      visible={true}
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "justifyContent": "center",
            },
            {
              "backgroundColor": "rgba(64,64,64,0.8)",
            },
            undefined,
          ]
        }
      >
        <View
          style={
            [
              {},
              undefined,
            ]
          }
        />
      </View>
    </Modal>
  </View>
  <View
    entering={BaseAnimationMock {}}
    exiting={BaseAnimationMock {}}
    layout={BaseAnimationMock {}}
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 10,
          "height": 100,
          "paddingHorizontal": 24,
          "width": "100%",
          "zIndex": 100,
        },
        {
          "alignSelf": "center",
          "top": 150,
        },
        undefined,
        {
          "backgroundColor": "transparent",
          "marginTop": -110,
          "position": "absolute",
        },
      ]
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "flex": 1,
          "flexDirection": "row",
          "opacity": 1,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 100,
            "height": 28,
            "justifyContent": "center",
            "marginLeft": 18,
            "marginTop": 21,
            "width": 28,
            "zIndex": 100,
          }
        }
      />
      <View
        style={
          {
            "flex": 1,
            "zIndex": 100,
          }
        }
      >
        <Text
          ellipsizeMode="middle"
          numberOfLines={3}
          style={
            [
              {
                "fontWeight": "500",
                "marginTop": 17,
                "paddingHorizontal": 20,
              },
              {
                "color": "#FFFFFF",
              },
            ]
          }
        />
        <Text
          ellipsizeMode="middle"
          numberOfLines={3}
          style={
            [
              {
                "fontWeight": "400",
                "marginTop": 6,
                "paddingHorizontal": 20,
              },
              {
                "color": "#FFFFFF",
              },
            ]
          }
        />
      </View>
    </View>
  </View>
</View>
`;
